### PR TITLE
Add CloudFormation template for bootstrap

### PR DESCRIPTION
**Pull request by randomvariable**
**What this PR does / why we need it**:

Adds a typed bootstrap CloudFormation template to set up IAM roles
and instance profiles that can be used by the Cluster API controllers
and the AWS Cloud Provider.

Also adds a `clusterawsadm` utility for want of a better name to
generate the YAML files and optionally apply it.

Commands are namespaced into `alpha`, as with `kubeadm` to not provide
a guarantee about its future availability.

This is part of improving user experience. Documentation updates to
follow in another PR.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #120 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cloud/aws/services/cloudformation/bootstrap.go [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+// IAM policy
+func CreateBootstrapStack() error {
+	stackName := "cluster-api-provider-aws-sigs-k8s-io"
+	sess := session.New()
+	svc := cfn.New(sess)
+
+	template := BootstrapTemplate()
+	yaml, err := template.YAML()
+	if err != nil {
+		return err
```
Comment (chuckha) [inline cloud/aws/services/cloudformation/bootstrap.go]: please wrap errors with `errors.Wrap(err, "failed to generate cloudformation YAML")` or similar message. This makes it much easier to debug both during development and a user submits a ticket.

File: cloud/aws/services/cloudformation/bootstrap.go [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+	}
+
+	input := &cfn.CreateStackInput{
+		Capabilities: aws.StringSlice([]string{"CAPABILITY_IAM", "CAPABILITY_NAMED_IAM"}),
```
Comment (chuckha) [inline cloud/aws/services/cloudformation/bootstrap.go]: these are both consts on the cfn sdk: `cfn.CapabilityCapabilityIam` and `cfn.CapabilityCapabilityNamedIam`

File: cloud/aws/services/cloudformation/bootstrap.go [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+		TemplateBody: aws.String(string(yaml)),
+		StackName:    aws.String(stackName),
+	}
+
+	_, err = svc.CreateStack(input)
+
+	glog.Infof("creating AWS CloudFormation stack %q", stackName)
```
Comment (chuckha) [inline cloud/aws/services/cloudformation/bootstrap.go]: the other services have been using `glog.V(2)` for logging that is reporting what resources have been modified

File: cmd/clusterawsadm/cmd/logutil.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/spf13/pflag"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+var logFlushFreq = pflag.Duration("log-flush-frequency", 5*time.Second, "Maximum number of seconds between log flushes")
+
+// TODO(thockin): This is temporary until we agree on log dirs and put those into each cmd.
```
Comment (chuckha) [inline cmd/clusterawsadm/cmd/logutil.go]: I'm curious about where this came from, is this a copy/pasta? might be good to add a comment with the source at the top, maybe also why we need it

File: cmd/clusterawsadm/cmd/root.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	"flag"
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"sigs.k8s.io/cluster-api-provider-aws/cmd/clusterawsadm/cmd/alpha"
+)
+
+var RootCmd = &cobra.Command{
```
Comment (chuckha) [inline cmd/clusterawsadm/cmd/root.go]: package level vars can be hard to work with and easily abused. It's ok to get get this moving right now, but can we track this as cleanup work in an issue if you agree?

Comment (randomvariable) [reply cmd/clusterawsadm/cmd/root.go]: Yup. This is straight up hackiness based on copying directly from the cluster-api repo.

File: cmd/clusterawsadm/cmd/logutil.go (hunk 4 above)
Comment (randomvariable) [reply cmd/clusterawsadm/cmd/logutil.go]: Yup, straight from Cluster API. I can imagine we would ditch glog in favour of logrus at some point?

Comment (randomvariable) [reply cmd/clusterawsadm/cmd/logutil.go]: Have added a "TODO" and link to source

File: cloud/aws/services/cloudformation/bootstrap.go [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+		return errors.Wrap(err, "failed to generate AWS CloudFormation YAML")
+	}
+
+	input := &cfn.CreateStackInput{
+		Capabilities: aws.StringSlice([]string{cfn.CapabilityCapabilityIam, cfn.CapabilityCapabilityNamedIam}),
+		TemplateBody: aws.String(string(yaml)),
+		StackName:    aws.String(stackName),
+	}
+
+	_, err = svc.CreateStack(input)
```
Comment (vincepri) [inline cloud/aws/services/cloudformation/bootstrap.go]: Looks like this error is being ignored, isn't it?

File: cloud/aws/services/cloudformation/bootstrap.go [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+
+	glog.V(2).Infof("creating AWS CloudFormation stack %q", stackName)
+	desInput := &cfn.DescribeStacksInput{StackName: aws.String(stackName)}
+	glog.V(2).Infof("waiting for stack %q to create", stackName)
+	err = svc.WaitUntilStackCreateComplete(desInput)
```
Comment (vincepri) [inline cloud/aws/services/cloudformation/bootstrap.go]: Might want to use `if err := svc.WaitUntilStackCreateComplete(desInput); err != nil` version to check errors and avoid overwriting the outer err variable

File: cloud/aws/services/iam/types.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+	EffectAllow = "Allow"
+
+	// EffectDeny is the Deny effect in an IAM policy statement entry
+	EffectDeny = "Deny"
+
+	// IAMSuffix is a standard prefix for resources for Cluster API Provider AWS
+	IAMSuffix = "cluster-api-provider-aws.sigs.k8s.io"
+
+	// PrincipalAws is the principal covering AWS arns.
+	PrincipalAws = "AWS"
```
Comment (vincepri) [inline cloud/aws/services/iam/types.go]: Should this be `PrincipalAWS`?

File: cloud/aws/services/iam/types.go [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+	ID        string `json:"id,omitempty"`
+}
+
+// StatementEntry represents each "statement" block in an IAM policy document
+type StatementEntry struct {
+	Sid          string     `json:",omitempty"`
+	Principal    Principals `json:",omitempty"`
+	NotPrincipal Principals `json:",omitempty"`
+	Effect       string
+	Action       Actions
```
Comment (vincepri) [inline cloud/aws/services/iam/types.go]: Should effect and action not be in the json?

File: cloud/aws/services/iam/types.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+type PrincipalID []string
+
+// Conditions is the map of all conditions in the statement entry.
+type Conditions map[string]ConditionValues
+
+// ConditionValues are a list of condition values in a condition statement
+type ConditionValues []string
+
+// JSON is the JSON output of the policy document
+func (p *PolicyDocument) JSON() (*[]byte, error) {
```
Comment (vincepri) [inline cloud/aws/services/iam/types.go]: Any reason the return value is a pointer to a slice of bytes?

File: cmd/clusterawsadm/cmd/root.go [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+	Use:   "clusterawsadm",
+	Short: "cluster api aws management",
+	Long:  `Cluster API Provider AWS commandst`,
```
Comment (vincepri) [inline cmd/clusterawsadm/cmd/root.go]: Typo at the end?

File: README.md [hunk 12]
```diff
@@ -107,6 +107,16 @@ After running clusterctl, a minikube instance is going to be running locally, wh
 - To tail cluster actuator logs, `kubectl get po -o name | grep clusterapi-controllers | xargs kubectl logs -c aws-machine-controller -f`.
 - To tail machine actuator logs, `kubectl get po -o name | grep clusterapi-controllers | xargs kubectl logs -c aws-cluster-controller -f`.
 
+#### CloudFormation
```
Comment (vincepri) [inline README.md]: Should we rephrase this to be something like `Setting up prerequisites with CloudFormation`? What do you think about adding a list of resources (with bullet points) that are needed?

Comment (randomvariable) [reply README.md]: #161 replaces this README.md, so I might as well delete it here

File: cloud/aws/services/cloudformation/bootstrap.go (hunk 6 above)
Comment (randomvariable) [reply cloud/aws/services/cloudformation/bootstrap.go]: Yup. Sorted, thanks.

File: cloud/aws/services/cloudformation/bootstrap.go (hunk 7 above)
Comment (randomvariable) [reply cloud/aws/services/cloudformation/bootstrap.go]: Changed.

File: cloud/aws/services/iam/types.go (hunk 8 above)
Comment (randomvariable) [reply cloud/aws/services/iam/types.go]: Yes, if matching AWS SDKs. Thanks.

File: cloud/aws/services/iam/types.go (hunk 9 above)
Comment (randomvariable) [reply cloud/aws/services/iam/types.go]: That was implicit, but have made explicit now.

File: cloud/aws/services/iam/types.go (hunk 10 above)
Comment (randomvariable) [reply cloud/aws/services/iam/types.go]: Nope. Changed to string.

File: cmd/clusterawsadm/cmd/root.go (hunk 11 above)
Comment (randomvariable) [reply cmd/clusterawsadm/cmd/root.go]: Thanks.

**Comment by randomvariable** [comment]
#161 has a getting started guide which explains how to use this.
